### PR TITLE
Add OpenAPI security (rebase)

### PR DESCRIPTION
**Pull request by gcasar**
Fixes [Issue #35](https://github.com/wI2L/fizz/issues/35#issue-492088752), continues stale PR #50.

I tried to keep the comments from #50 relevant by resolving conflicts by back-merging and addressing them here. We can squash commits at your discretion.

- adds generator methods
- some edge case test coverage
- marshalling empty `security` field in path section (see [security](https://swagger.io/specification/#oauth-flows-object) parameter - an empty array is used to signal "no security" override)

The last point is the only one that is troublesome since it does not align with the way JSON gets serialised and would benefit from something like [this request](https://github.com/golang/go/issues/22480). The proposed way is backwards compatible but does repeat some code.

### CODE REVIEW

**Comment by wI2L** [comment]
@gcasar Thanks for picking this up. I'll review it asap this week.

Regarding the `omitnil` proposal you linked, you could check out my other package `jettison`, but after a quick peek to your custom marshaller implem, I think it's better than introducing a new dependency just for that need (which would only be for JSON marshaling anyway, and not YAML).

File: openapi/spec.go [hunk 1]
```diff
@@ -184,6 +188,24 @@ type Schema struct {
...
+	RequestBody  *RequestBody           `json:"requestBody,omitempty" yaml:"requestBody,omitempty"`
+	Responses    Responses              `json:"responses,omitempty" yaml:"responses,omitempty"`
+	Deprecated   bool                   `json:"deprecated,omitempty" yaml:"deprecated,omitempty"`
+	Servers      []*Server              `json:"servers,omitempty" yaml:"servers,omitempty"`
+	Security     []*SecurityRequirement `json:"security" yaml:"security"`
+	XCodeSamples []*XCodeSample         `json:"x-codeSamples,omitempty" yaml:"x-codeSamples,omitempty"`
+}
+
+// A workaround for missing omitnil functionality.
+// explicitly excludes Security which:
```
Comment (wI2L) [inline openapi/spec.go:205]: I would rephrase the comment to smthing like this, up to you tho:
"It explicitely omit the Security field from marshaling when it is nil, but not when empty."

File: openapi/generator.go [hunk 2]
```diff
@@ -104,6 +104,12 @@ func (g *Generator) SetServers(servers []*Server) {
 	g.api.Servers = servers
 }
 
+// SetSecurity sets the security options for the
+// current specification.
+func (g *Generator) SetSecurity(security *SecurityRequirement) {
```
Comment (wI2L) [inline openapi/generator.go:109]: I would prefer to name the function `SetSecurityRequirement`, to be in-par with the OpenAPI spec object naming.

File: fizz.go [hunk 3]
```diff
@@ -353,6 +353,20 @@ func InputModel(model interface{}) func(*openapi.OperationInfo) {
...
 
+// Security adds a security definition to an operation.
+func Security(security *openapi.SecurityRequirement) func(*openapi.OperationInfo) {
+	return func(o *openapi.OperationInfo) {
+		o.Security = append(o.Security, security)
+	}
+}
+
+// NoSecurity explicitly removes all security.
+func NoSecurity() func(*openapi.OperationInfo) {
```
Comment (wI2L) [inline fizz.go:364]: Having this feels weird for the end-user. I'm wondering if it shouldn't be the default, applied directly by the package (i.e, if `operation.Security == nil`, initialize `Security` to empty slice, via `MarshalYAML/JSON`) ?

**Review by wI2L — CHANGES_REQUESTED** [review]

File: fizz.go [hunk 4]
```diff
@@ -353,6 +353,20 @@ func InputModel(model interface{}) func(*openapi.OperationInfo) {
 	}
 }
 
+// Security adds a security definition to an operation.
+func Security(security *openapi.SecurityRequirement) func(*openapi.OperationInfo) {
```
Comment (wI2L) [inline fizz.go:357]: Please add this to the README: https://github.com/wI2L/fizz#operation-informations

File: fizz.go (hunk 3 above)
Comment (gcasar) [reply fizz.go:364]: Hmm... given that I understand your comment correctly I believe that would actually change semantics:
- you want to omit Security altogether (as per the spec) if you are not overriding the top level Security setting
- any serialized Security value, even an empty array has meaning.

From Operation Object:
> A declaration of which security mechanisms can be used for this operation. The list of values includes alternative security requirement objects that can be used. Only one of the security requirement objects need to be satisfied to authorize a request. To make security optional, an empty security requirement ({}) can be included in the array. This definition overrides any declared top-level security. **To remove a top-level security declaration, an empty array can be used.**

Comment (wI2L) [reply fizz.go:364]: Okay, that was me, I didn't read the spec properly. `NoSecurity()` would be useful only if the top-level `Security` field is set:

> To remove a top-level security declaration, an empty array can be used.

What would you think of `WithoutSecurity` instead as name ?

---

Speaking of that, the spec also specify the following:

>  To make security optional, an empty security requirement ({}) can be included in the array.

We could provide a "helper-func" (`WithOptionalSecurity()`) for that, that would append an empty `openapi.SecurityRequirement` object in the operation's `Security` field (slice). Thoughts ?

**Comment by wI2L** [comment]
@gcasar Sorry for the delay, just got back from holidays. Please see the minor fixes suggestions, otherwise looks good.
I'd rather squash indeed, for commit logs brevity.

**Comment by wI2L** [comment]
Closes #50 if merged.

File: fizz.go (hunk 4 above)
Comment (wI2L) [reply fizz.go:357]: Also, if we're including security-related helpers functions such as `WithoutSecurity()` and `OptionalSecurity()`, they should also be documented/mentionned.

File: fizz.go (hunk 3 above)
Comment (gcasar) [reply fizz.go:364]: I like it 👍

File: README.md [hunk 5]
```diff
@@ -96,6 +96,17 @@ fizz.Header(name, desc string, model interface{})
 // Override the binding model of the operation.
 fizz.InputModel(model interface{})
 
+// Add a security requirement to the list of security overrides for this operation.
+// Note that this removes all other SecurityRequirements
+// that might have been specified in the root `security` property.
+fizz.Security(security *openapi.SecurityRequirement)
+
+// Make other security requirements optional.
```
Comment (wI2L) [inline README.md:104]: Again, I think it is better to be explicit. We should indicate what the function does, not just its effect:

> Add an empty security requirement to this operation to make other security requirements optional.

File: README.md [hunk 6]
```diff
@@ -96,6 +96,17 @@ fizz.Header(name, desc string, model interface{})
...
+fizz.WithOptionalSecurity()
+
+// Remove any security requirements for this operation.
```
Comment (wI2L) [inline README.md:107]: > Remove any top-level security requirements for this operation.

File: fizz.go [hunk 7]
```diff
@@ -353,6 +353,28 @@ func InputModel(model interface{}) func(*openapi.OperationInfo) {
...
+
+// WithOptionalSecurity makes other security requirements optional.
+func WithOptionalSecurity() func(*openapi.OperationInfo) {
+	return func(o *openapi.OperationInfo) {
+		var emptyRequirement openapi.SecurityRequirement = make(openapi.SecurityRequirement)
+		o.Security = append(o.Security, &emptyRequirement)
+	}
+}
+
+// WithoutSecurity explicitly removes all security.
```
Comment (wI2L) [inline fizz.go:371]: Use same comment as those in README. See proposal.

File: fizz.go [hunk 8]
```diff
@@ -353,6 +353,28 @@ func InputModel(model interface{}) func(*openapi.OperationInfo) {
...
 }
 
+// Security adds a security definition to an operation.
+func Security(security *openapi.SecurityRequirement) func(*openapi.OperationInfo) {
+	return func(o *openapi.OperationInfo) {
+		o.Security = append(o.Security, security)
+	}
+}
+
+// WithOptionalSecurity makes other security requirements optional.
```
Comment (wI2L) [inline fizz.go:363]: Use same comment as those in README. See proposal.

**Review by wI2L — CHANGES_REQUESTED** [review]
Thanks.
Few docs fixes proposed, but otherwise LGTM.
Please squash your commits, and I'll merge it ASAP (might take some time before to fix the CI/CD workflow to switch from Jenkins to GitHub Actions).

File: README.md (hunk 5 above)
Comment (gcasar) [reply README.md]: I was not sure how to break this one - let me know if you want me to reformat it:)

**Comment by gcasar** [comment]
@wI2L almost there :) I don't like to squash by force pushing - is it ok with you if you squash it yourself when merging?

**Comment by ipfans** [comment]
Any updates @wI2L ?

**Comment by wI2L** [comment]
@ipfans will take care of migrating CI to Github actions and merge pending pull requests to make a proper release by september. Sorry for the delay.

**Comment by nikicc** [comment]
Any updates on this? I'd find it useful 😊

**Review by wI2L — DISMISSED** [review]

**Review by wI2L — APPROVED** [review]

**Comment by wI2L** [comment]
To everyone involved on this PR, and expecting the changes, I apologize for the unreasonable delay, haven't found some time to work on it, and I wanted to update the CI pipeline of the project before merhing anything new. This will be shipped in next release.

**Comment by nikicc** [comment]
Thanks for getting it in @wI2L!
